MSAL/src/util/ios/MSALRedirectUriVerifier.m

```diff
@@ -26,63 +26,111 @@
 
-        if (![host isEqualToString:bundleIdentifier])
+        if (![self verifySchemeIsRegistered:redirectURI error:error])
         {
-            MSAL_ERROR_PARAM(nil, MSALErrorRedirectSchemeNotRegistered, @"Your MSALPublicClientApplication is configured to allow brokered authentication but your redirect URI is not setup properly. Make sure your redirect URI is in the form of <app-scheme>://<bundle-id> (e.g. \"x-msauth-testapp://com.microsoft.msal.testapp\") and that the \"app-scheme\" you choose is registered in your application's info.plist.");
-
-            return NO;
+            return nil;
         }
+
+        BOOL brokerCapable = [self redirectUriIsBrokerCapable:redirectURI];
+
+        MSALRedirectUri *redirectUri = [[MSALRedirectUri alloc] initWithRedirectUri:redirectURI
+                                                                      brokerCapable:brokerCapable];
+
+        return redirectUri;
     }
 
-    return YES;
-}
+    // First try to check for broker capable redirect URI
+    NSURL *defaultRedirectUri = [self defaultBrokerCapableRedirectUri];
 
-+ (NSURL *)generateRedirectUri:(NSString *)inputRedirectUri
-                      clientId:(NSString *)clientId
-                 brokerEnabled:(BOOL)brokerEnabled
-                         error:(NSError **)error
-{
-    if (![NSString msidIsStringNilOrBlank:inputRedirectUri])
+    NSError *redirectError = nil;
+    if ([self verifySchemeIsRegistered:defaultRedirectUri error:&redirectError])
     {
-        return [NSURL URLWithString:inputRedirectUri];
+        return [[MSALRedirectUri alloc] initWithRedirectUri:defaultRedirectUri brokerCapable:YES];
     }
 
+    // Now check the uri that is not broker capable for backward compat
+    defaultRedirectUri = [self defaultNonBrokerRedirectUri:clientId];
+
+    if ([self verifySchemeIsRegistered:defaultRedirectUri error:nil])
+    {
+        return [[MSALRedirectUri alloc] initWithRedirectUri:defaultRedirectUri brokerCapable:NO];
+    }
+
+    if (error)
+    {
+        *error = redirectError;
+    }
+
+    return nil;
+}
+
++ (NSURL *)defaultNonBrokerRedirectUri:(NSString *)clientId
+{
     if ([NSString msidIsStringNilOrBlank:clientId])
     {
-        MSAL_ERROR_PARAM(nil, MSALErrorInternal, @"The client ID provided is empty or nil.");
         return nil;
     }
 
-    NSString *scheme = [NSString stringWithFormat:@"msal%@", clientId];
-    NSString *hostComponent = brokerEnabled ? [[NSBundle mainBundle] bundleIdentifier] : @"auth";
-    NSString *redirectUriString = [NSString stringWithFormat:@"%@://%@", scheme, hostComponent];
-    return [NSURL URLWithString:redirectUriString];
+    NSString *redirectUri = [NSString stringWithFormat:@"msal%@://auth", clientId];
+    return [NSURL URLWithString:redirectUri];
+}
+
++ (NSURL *)defaultBrokerCapableRedirectUri
+{
+    NSString *bundleID = [[NSBundle mainBundle] bundleIdentifier];
+    NSString *redirectUri = [NSString stringWithFormat:@"msauth%@://auth", bundleID];
```

Comment:
nit: do we want to use the constant added in msid?

Reply:
and msauthv2, no?

Reply:
Constants added in msid are for the broker scheme. This is MSAL redirect uri prefix. I'd like to keep those 2 separate. 

I'm also not fully sure if it should have v2 or not, as broker scheme is something more internal and this URL will be shown to everyone as the default MSAL redirect uri. Thoughts?